pubtools/_pulp/tasks/copy_repo.py

```diff
@@ -0,0 +1,324 @@
+            f = f_map(f, partial(RepoCopy, repo=dest_repo))
+            f = f_map(f, self.log_copy)
+            futures.append(f)
+
+        return futures
+
+    @step("Record push items")
+    def record_copies(self, copied_repos_fs):
+        return [f_flat_map(f, self.record_copied_repo) for f in copied_repos_fs]
+
+    def record_copied_repo(self, copied_repo):
+        push_items = []
+        for task in copied_repo.tasks:
+            push_items.extend(self.push_items_for_task(task))
+        return self.collector.update_push_items(push_items)
+
+    def push_items_for_task(self, task):
+        out = []
+        for unit in task.units:
+            push_item = self.push_item_for_unit(unit)
+            if push_item:
+                out.append(push_item)
+        return out
+
+    def push_item_for_unit(self, unit):
+        for unit_type, fn in [
+            (ModulemdUnit, self.push_item_for_modulemd),
+            (RpmUnit, self.push_item_for_rpm),
+            (FileUnit, self.push_item_for_file),
+        ]:
+            if isinstance(unit, unit_type):
+                return fn(unit)
+
+    def push_item_for_modulemd(self, unit):
+        out = {}
+        out["state"] = "PUSHED"
+        out["origin"] = "pulp"
+
+        # Note: N:S:V:C:A format here is kept even if some part
+        # of the data is missing (never expected to happen).
+        # For example, if C was missing, you'll get N:S:V::A
+        # so the arch part can't be misinterpreted as context.
+        nsvca = ":".join(
+            [unit.name, unit.stream, str(unit.version), unit.context, unit.arch]
+        )
+
+        out["name"] = nsvca
+
+        return ModuleMdPushItem(**out)
+
+    def push_item_for_rpm(self, unit):
+        out = {}
+
+        out["state"] = "PUSHED"
+        out["origin"] = "pulp"
+
+        filename_parts = [
+            unit.name,
+            "-",
+            unit.version,
+            "-",
+            unit.release,
+            ".",
+            unit.arch,
+            ".rpm",
+        ]
+        out["name"] = "".join(filename_parts)
+
+        # Note: in practice we don't necessarily expect to get all of these
+        # attributes, as after a delete the server will only provide those
+        # which make up the unit key. We still copy them anyway (even if
+        # values are None) in case this is improved some day.
+        out["sha256sum"] = unit.sha256sum
+        out["md5sum"] = unit.md5sum
+        out["signing_key"] = unit.signing_key
+
+        return RpmPushItem(**out)
+
+    def push_item_for_file(self, unit):
+        out = {}
+
+        out["state"] = "PUSHED"
+        out["origin"] = "pulp"
+        out["name"] = unit.path
+        out["sha256sum"] = unit.sha256sum
+
+        return FilePushItem(**out)
+
+    @step("Publish")
+    def publish(self, repo_fs):
+        return [
+            f_flat_map(f, lambda r: r.publish(PublishOptions(clean=True)))
```

Comment:
I think you can remove clean=True, that would be specific to the "delete" use-case, I can't think of any reason it's needed after a copy.